### PR TITLE
Blog app: Process data in models

### PR DESCRIPTION
**Pull request by ernestmusong**
In this PR, I've done the following:

- Create two users named **first_user** and **second_user**
- Create 4 posts written by **first_user**
- Create 6 posts comments for the post named **first_post**

**In Users**
  -- Add a method that returns the 3 most recent posts for a given user.
**In Posts**
  -- Added a method that updates the posts counter for a user.
  -- Added a method which returns the 5 most recent comments for a given post.
**In Comments**
  -- Added a method that updates the comments counter for a post.
**In Likes**
  -- Added a method that updates the likes counter for a post
  
  ## check if methods are working.
**Go to rails c: Run the code below**

**1. Recent posts:**

user = User.find(1)
recent_posts = user.recent_posts
recent_posts.each do |post|
  puts "Title: #{post.title}"
  puts "Text: #{post.text}"
end

**2. Recent comments:**

post = Post.find(1)
recent_comments = post.recent_comments
recent_comments.each do |comment|
  puts "Comment: #{comment.text}"
end
 
**3. Update comments counter for a post:**

comment = Comment.find(1)
comment.update_post_comments_counter 

**4. Update the likes counter for a post.**

like = Like.find(1)
like.update_post_likes_counter

### CODE REVIEW

File: app/models/post.rb [hunk 1]
```diff
@@ -0,0 +1,13 @@
+class Post < ApplicationRecord
+  belongs_to :author, class_name: 'User', foreign_key: :author_id
```
Comment (rloterh) [inline app/models/post.rb]: - [ ] The `id` of _Post_ is supposed to reference the foreign key `author_id` of the _User_ model (and also declared in the _User_ model), and hence it's redundant to declare `author_id` as a foreign key in _Post_ model. Simply do something like:
```ruby
# app/models/post.rb
belongs_to :author, class_name: 'User'

File: app/models/comment.rb [hunk 2]
```diff
@@ -0,0 +1,8 @@
+class Comment < ApplicationRecord
+  belongs_to :author, class_name: 'User', foreign_key: :author_id
```
Comment (rloterh) [inline app/models/comment.rb]: - [ ] Similarly, it's redundant to declare `author_id` as a foreign key in _Comment_ model as The `id` of _Comment_ references the foreign key `author_id` of the _User_ model (and also declared in the _User_ model), and hence. Simply do something like:
```ruby
# app/models/comment.rb
belongs_to :author, class_name: 'User'

File: app/models/like.rb [hunk 3]
```diff
@@ -0,0 +1,8 @@
+class Like < ApplicationRecord
+  belongs_to :author, class_name: 'User', foreign_key: :author_id
```
Comment (rloterh) [inline app/models/like.rb]: - [ ] Again, it's redundant to declare `author_id` as a foreign key in _Like_ model as The `id` of _Like_ references the foreign key `author_id` of the _User_ model (and also declared in the _User_ model), and hence. Simply do something like:
```ruby
# app/models/Like.rb
belongs_to :author, class_name: 'User'

File: app/models/post.rb [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+  has_many :comments
+  has_many :likes
+
+  def update_user_posts_counter
+    author.increment!(:posts_counter)
+  end
```
Comment (rloterh) [inline app/models/post.rb]: - [ ] [OPTIONAL]
It's a good practice to use one of the Rails Active Callbacks `after_save` to call the `update_post_counter` method every time a new post has been saved to the database, and when `after_save` is called mark the `update_post_counter` method as _private_ accordingly, which should be similar to:

```ruby
 # app/models/post.rb
  after_save :update_post_counter

  private

  def update_post_counter
    author.increment!(:posts_counter)
  end
```

File: app/models/comment.rb [hunk 5]
```diff
@@ -0,0 +1,8 @@
...
+  belongs_to :post
+
+  def update_post_comments_counter
+    post.update(comments_counter: post.comments.count)
+  end
```
Comment (rloterh) [inline app/models/comment.rb:8]: - [ ] [OPTIONAL]
Similarly, it's a good practice to use one of the Rails Active Callbacks `after_save` to call the `update_comment_counter` method every time a new comment has been saved to the database, and when `after_save` is called mark the `update_comment_counter` method as _private_ accordingly, which should be similar to:

```ruby
 # app/models/comment.rb
  after_save :update_comment_counter

  private

  def update_comment_counter
    author.increment!(:comments_counter)
  end
```

File: app/models/like.rb [hunk 6]
```diff
@@ -0,0 +1,8 @@
...
+  belongs_to :post
+
+  def update_post_likes_counter
+    post.update(likes_counter: post.likes.count)
+  end
```
Comment (rloterh) [inline app/models/like.rb:10]: - [ ] [OPTIONAL]
Again, it's a good practice to use one of the Rails Active Callbacks `after_save` to call the `update_like_counter` method every time a new like has been saved to the database, and when `after_save` is called mark the `update_like_counter` method as _private_ accordingly, which should be similar to:

```ruby
 # app/models/like.rb
  after_save :update_like_counter

  private

  def update_like_counter
    author.increment!(:likes_counter)
  end
```

**Review by rloterh — CHANGES_REQUESTED** [review]
Hi Musong, 

You did extremely great on this project! 👏 👏 
There are however a few issues that you still need to work on to go to the next project but you are almost there! 
🏁 🚦  🚗 _ 🏃 🌴  🏠 🌳 🏘️ 🏡  

## Highlights 👍  
- Good job so far with creating model associations!
- ReadMe looks great!

## Required Changes ♻️
_Kindly check the inline comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you consider them as they can improve your code._

_Kindly consider my inline comments marked **[OPTIONAL]** under this review_

Cheers and Happy coding!👏 💻 🍷

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think the code review was unfair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by ernestmusong** [comment]
Thank you @rloterh for your feedback. I have implemented the changes as required above.

**Review by Olamarx — APPROVED** [review]
Hi @ernestmusong, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## Highlights
- All required changes fixed ✔️
- Association setup ✔️
- Linters passed ✔️
- Readme file ✔️

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**